api/src/main/java/org/jenkinsci/plugins/workflow/flow/GraphListener.java

```diff
@@ -38,4 +38,12 @@
      * One of the use cases of this listener is to persist the state of {@link FlowExecution}.
      */
     void onNewHead(FlowNode node);
+
+    /**
+     * Listener which should be notified of events immediately as they occur.
+     * You must be very careful not to acquire locks or block.
+     * If you do not implement this marker interface, you will receive notifications in batched deliveries.
+     */
+    interface Synchronous extends GraphListener {}
```

Comment:
Do not follow. The existing behavior was asynchronous listeners. I just added an option for synchronous listeners. From what I know of #285 it should continue to use asynchronous listeners so this is irrelevant.